include/revng/Pipeline/ReadPathsCache.h

```diff
@@ -0,0 +1,215 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <tuple>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Pipeline/Target.h"
+#include "revng/TupleTree/TupleTreePath.h"
+
+namespace pipeline {
+class Context;
+
+class ParsableLocatedTarget {
+public:
+  std::string Content;
+  std::string StepName;
+  std::string ContainerName;
+};
+
+class LocatedTarget {
+private:
+  Target Content;
+  std::string StepName;
+  std::string ContainerName;
+
+public:
+  LocatedTarget(const Target &Target,
+                const std::string &StepName,
+                const std::string &ContainerName) :
+    Content(Target), StepName(StepName), ContainerName(ContainerName) {}
+
+  static llvm::Expected<llvm::SmallVector<LocatedTarget, 2>>
+  fromParsable(const Context &Ctx, const ParsableLocatedTarget &Target);
+
+  const Target &getTarget() const { return Content; }
+
+  llvm::StringRef getStepName() const { return llvm::StringRef(StepName); }
+
+  llvm::StringRef getContainerName() const {
+    return llvm::StringRef(ContainerName);
+  }
+
+  Target &getTarget() { return Content; }
+
+  std::string &getStepName() { return StepName; }
+
+  std::string &getContainerName() { return ContainerName; }
+
+  bool operator==(const LocatedTarget &) const = default;
+
+  bool operator!=(const LocatedTarget &) const = default;
+
+  bool operator<(const LocatedTarget &Other) const {
+    const auto &LHS = std::tie(Content, StepName, ContainerName);
+    return LHS < std::tie(Other.Content, Other.StepName, Other.ContainerName);
+  }
+
+  ParsableLocatedTarget toParsable() const;
+};
+
+class TargetReadPathsSerializable {
+public:
+  std::string Path;
+};
+
+class PathSerializationPrimitives {
+public:
+  PathSerializationPrimitives() = default;
+
+  template<TupleTreeCompatibleAndVerifiable T>
+  static PathSerializationPrimitives make() {
+    PathSerializationPrimitives Primitives;
+    Primitives.Deserialize = &stringAsPath<T>;
+    Primitives.Serialize = &pathAsString<T>;
+    return Primitives;
+  }
+
+  std::optional<TupleTreePath> (*Deserialize)(llvm::StringRef);
+  std::optional<std::string> (*Serialize)(const TupleTreePath &);
+};
+
+class TargetReadPaths {
+private:
+  TupleTreePath Path;
```

Comment:
1. Why is this class called `TargetReadPaths` (plural) if it holds a single path?
2. Can we drop this class in favor of a plain `TupleTreePath` plus a pair of free functions to serialize/deserialize?
   Maybe instead of free functions these could be methods of `Context` or `Global`?
   There's quite a few classes with non-trivial names but very simple content that make decoding this file rather difficult.

Reply:
i entirely removed the class